src/flint/types/fq_default.pyx

```diff
@@ -0,0 +1,886 @@
+        if typecheck(obj, nmod):
+            fq_default_set_ui(fq_ele, <ulong>(<nmod>obj).val, self.val)
+            return 0
+
+        # For fmpz_mod we can also convert directly
+        # Assumes that the modulus of the ring matches the context
+        if typecheck(obj, fmpz_mod):
+            fq_default_set_fmpz(fq_ele, (<fmpz_mod>obj).val, self.val)
+            return 0
+
+        # Otherwise the object wasn't a scalar we convert from
+        return NotImplemented
+
+    cdef set_any_as_fq_default(self, fq_default_t fq_ele, obj):
+        # First try and convert from scalars
+        check = self.set_any_scalar_as_fq_default(fq_ele, obj)
+        if check is not NotImplemented:
+            return 0
+
+        # Assumes that the modulus of the polynomial matches
+        # the context for the fq_default
+        if typecheck(obj, fmpz_mod_poly):
+            fq_default_set_fmpz_mod_poly(fq_ele, (<fmpz_mod_poly>obj).val, self.val)
+            return 0
+
+        # Assumes that the modulus of the polynomial matches
+        # the context for the fq_default
+        if typecheck(obj, nmod_poly):
+            fq_default_set_nmod_poly(fq_ele, (<nmod_poly>obj).val, self.val)
+            return 0
+
+        # If the input is not fmpz_mod_poly or nmod_poly or a list, we cast the
+        # input to an fmpz_poly and then set from this
+        poly = any_as_fmpz_poly(obj)
+        if poly is NotImplemented:
+            return NotImplemented
+
+        fq_default_set_fmpz_poly(fq_ele, (<fmpz_poly>poly).val, self.val)
+        return 0
+
+    cdef any_as_fq_default(self, obj):
+        # convert from fq_default
+        if typecheck(obj, fq_default):
+            if self != (<fq_default>obj).ctx:
+                raise ValueError("contexts dont match")
+            return obj
+
+        cdef fq_default res
+        res = self.new_ctype_fq_default()
+        check = self.set_any_as_fq_default(res.val, obj)
+        if check is NotImplemented:
+            return NotImplemented
+        return res
+
+    def __eq__(self, other):
+        """
+        Two finite field context compare equal if they have same
+        characteristic, modulus, type and variable
+
+            >>> from flint import fmpz_mod_poly_ctx
+            >>> modulus = fmpz_mod_poly_ctx(5)([2,4,1])
+            >>> gf = fq_default_ctx(5, 2)
+            >>> gf2 = fq_default_ctx(modulus=modulus)
+            >>> gf2 == gf
+            True
+            >>> gf3 = fq_default_ctx(modulus=modulus, var="y")
+            >>> gf3 == gf
+            False
+        """
+        if self is other:
+            return True
+
+        if typecheck(other, fq_default_ctx):
+            return (self.fq_type == other.fq_type
+                    and self.var == other.var
+                    and self.prime() == other.prime()
+                    and self.modulus() == other.modulus())
+        else:
+            raise TypeError(f"Cannot compare fq_default_ctx with {type(other)}")
```

Comment:
`__eq__` should return `NotImplemented` rather than raising `TypeError`:
```
In [67]: F == None
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[67], line 1
----> 1 F == None

File fq_default.pyx:451, in flint.types.fq_default.fq_default_ctx.__eq__()

TypeError: Cannot compare fq_default_ctx with <class 'NoneType'>
```
Generally considered a faux-pas for `==` to raise an exception rather than turning into `False`.